File: ios/Kinowo/Models/DateLabel.swift
import Foundation

/// Formats a `DayShowings.date` (`YYYY-MM-DD`) into the long weekday + day +
/// month label the web renders for the same date (`DateFormatter.format` in
/// `web/src/main/scala/controllers/DateFormatter.scala`), e.g.
/// "Czwartek 4 czerwca" / "Thursday 4 June" — in the CALLER'S locale, never
/// baked server-side.
///
/// `DayShowings.label` used to be shown directly, but it's rendered ONCE by
/// the server in the deployment's fixed default language and never changes
/// when the visitor picks a different in-app language (`LanguageSelection`,
/// independent of the deployment/country) — the weekday/month text stayed
/// stuck in the old language after a switch while every other string on
/// screen followed the pick. Deriving it here, from the raw `date`, in
/// whatever `Locale` the caller passes (typically `@Environment(\.locale)`,
/// which `KinowoApp` keeps in step with the pick) fixes that the same way
/// the web's client-side `formatDateLabel` (`shared.js`) does.
enum DateLabel {
    private static let polishDays = [
        "poniedziałek", "wtorek", "środa", "czwartek", "piątek", "sobota", "niedziela",
    ]
    private static let polishMonths = [
        "", "stycznia", "lutego", "marca", "kwietnia", "maja", "czerwca",
        "lipca", "sierpnia", "września", "października", "listopada", "grudnia",
    ]

    private static let isoCalendar: Calendar = {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        return calendar
    }()

    /// `isoDate` is `YYYY-MM-DD`. Returns `isoDate` itself if it doesn't parse
    /// (defensive only — every `date` this app decodes comes from the API's
    /// own ISO-formatted field).
    static func format(isoDate: String, locale: Locale) -> String {
        let parts = isoDate.split(separator: "-").compactMap { Int($0) }
        guard parts.count == 3,
              let date = isoCalendar.date(from: DateComponents(year: parts[0], month: parts[1], day: parts[2]))
        else { return isoDate }

        let (year, month, day) = (parts[0], parts[1], parts[2])
        let currentYear = isoCalendar.component(.year, from: Date())
        let yearSuffix  = year == currentYear ? "" : " \(year)"

        // `locale.language.languageCode` isn't available on Linux's
        // swift-corelibs-foundation — KinowoCore's `swift test` runs there in
        // CI's mobile-local-server job (Apple-only Foundation APIs compile
        // fine on local macOS `swift test`/`xcodebuild` but fail there; see
        // the Linux-Foundation-gap memory). Every `Locale` this app ever
        // builds is a bare 2-letter code (`LanguageSelection.supported`:
        // "pl"/"en"/"de"/"es"), so a plain prefix is both portable and
        // sufficient — no platform-specific API needed at all.
        let languageCode = locale.identifier.prefix(2).lowercased()
        let dayName: String
        let monthName: String
        if languageCode == "pl" {
            // Sunday=1…Saturday=7 → Monday-first index into `polishDays`.
            let mondayFirstIndex = (isoCalendar.component(.weekday, from: date) + 5) % 7
            dayName   = polishDays[mondayFirstIndex].capitalized
            monthName = polishMonths[month]
        } else {
            let formatter = Foundation.DateFormatter()
            formatter.locale = locale
            dayName   = formatter.weekdaySymbols[isoCalendar.component(.weekday, from: date) - 1].capitalized
            monthName = formatter.monthSymbols[month - 1]
        }
        return "\(dayName) \(day) \(monthName)\(yearSuffix)"
    }
}

File: android/app/src/main/java/pl/kinowo/model/DateLabel.kt
package pl.kinowo.model

import java.time.LocalDate
import java.time.format.TextStyle
import java.util.Locale

/**
 * Formats a [DayShowings.date] (`YYYY-MM-DD`) into the long weekday + day +
 * month label the web renders for the same date (`DateFormatter.format` in
 * `web/src/main/scala/controllers/DateFormatter.scala`), e.g. "Czwartek 4
 * czerwca" / "Thursday 4 June" — in the CURRENT locale, never
 * [DayShowings.label].
 *
 * [DayShowings.label] used to be shown directly, but it's rendered ONCE by
 * the server in the deployment's fixed default language and stays that way
 * regardless of the visitor's own in-app language pick
 * (`UserPreferences.selectedLanguageTag`, independent of the selected
 * country) — `recreate()` re-applies the picked locale to every OTHER string
 * on screen (`LocaleWrapper` forces `Locale.getDefault()`), but the already-
 * fetched `label` field doesn't follow it. Deriving it here from the raw
 * `date`, against [Locale.getDefault] at call time, fixes that the same way
 * the web's client-side `formatDateLabel` (`shared.js`) does.
 */
object DateLabel {
    private val polishDays = listOf(
        "poniedziałek", "wtorek", "środa", "czwartek", "piątek", "sobota", "niedziela",
    )
    private val polishMonths = listOf(
        "", "stycznia", "lutego", "marca", "kwietnia", "maja", "czerwca",
        "lipca", "sierpnia", "września", "października", "listopada", "grudnia",
    )

    /** [isoDate] is `YYYY-MM-DD`. Returns [isoDate] itself if it doesn't parse
     *  (defensive only — every `date` this app decodes comes from the API's
     *  own ISO-formatted field). */
    fun format(isoDate: String, locale: Locale = Locale.getDefault()): String {
        val date = runCatching { LocalDate.parse(isoDate) }.getOrNull() ?: return isoDate
        val currentYear = LocalDate.now().year
        val yearSuffix = if (date.year == currentYear) "" else " ${date.year}"
        val dayName = if (locale.language == "pl") {
            polishDays[date.dayOfWeek.value - 1].replaceFirstChar { it.uppercase() }
        } else {
            date.dayOfWeek.getDisplayName(TextStyle.FULL, locale).replaceFirstChar { it.uppercase() }
        }
        val monthName = if (locale.language == "pl") {
            polishMonths[date.monthValue]
        } else {
            date.month.getDisplayName(TextStyle.FULL, locale)
        }
        return "$dayName ${date.dayOfMonth} $monthName$yearSuffix"
    }
}
